File: cpp/fib.cpp
/*
  fib.cpp
  Minimalbeispiel zur Fibonacci-Berechnung in C++.
  Microsoft Visual C++ Compiler (MSVC) oder MinGW/Clang.
*/

#include <iostream>

int fib_rec(int n) {
    if (n <= 1) return n;
    return fib_rec(n - 1) + fib_rec(n - 2);
}

int fib_iter(int n) {
    if (n <= 1) return n;
    int a = 0, b = 1;
    for (int i = 2; i <= n; ++i) {
        int temp = a + b;
        a = b;
        b = temp;
    }
    return b;
}

int main() {
    int n;
    std::cout << "Gib n ein: ";
    std::cin >> n;

    std::cout << "Fibonacci iterativ(" << n << ") = " << fib_iter(n) << std::endl;
    std::cout << "Fibonacci rekursiv(" << n << ") = " << fib_rec(n) << std::endl;
    return 0;
}

File: rust/fib.rs
// fib.rs
//
// Kompilieren: rustc fib.rs -o fib
// Ausführen:   ./fib
//
// Zeigt, wie man in Rust Traits nutzt, um verschiedene Implementierungen
// der Fibonacci-Berechnung (rekursiv vs. iterativ) auszutauschen.

use std::io;

// Ein Trait "FibCalc" fungiert wie ein Interface:
// Wer dieses Trait implementiert, muss eine fib()-Methode anbieten.
trait FibCalc {
    fn fib(&self, n: u32) -> u32;
}

// Zwei Strukturen, die das Trait unterschiedlich implementieren:
struct RecFib;  // rekursiv
struct IterFib; // iterativ

impl FibCalc for RecFib {
    fn fib(&self, n: u32) -> u32 {
        if n <= 1 {
            n
        } else {
            self.fib(n - 1) + self.fib(n - 2)
        }
    }
}

impl FibCalc for IterFib {
    fn fib(&self, n: u32) -> u32 {
        if n <= 1 {
            return n;
        }
        let (mut a, mut b) = (0, 1);
        for _ in 2..=n {
            let temp = a + b;
            a = b;
            b = temp;
        }
        b
    }
}

fn main() {
    println!("Gib eine Zahl n ein:");
    let mut input = String::new();
    io::stdin().read_line(&mut input).unwrap();
    let n: u32 = input.trim().parse().unwrap();

    let recursive_calc = RecFib;
    let iterative_calc = IterFib;

    println!("IterFib: fib({}) = {}", n, iterative_calc.fib(n));
    println!("RecFib: fib({}) = {}", n, recursive_calc.fib(n));
}
